index.ts

```diff
@@ -0,0 +1,187 @@
+interface User2 {
+       id: number,
+       name: string,
+};
+
+// "Наследуясь" от предыдущих типов User1 и User2 создать новые аналогичные тип и интерфейс у которых помимо id и name 
+// будет еще свойство isAdmin с типом true
+type Admin1 = User1 & { isAdmin: true };
+interface Admin2 extends User2 {
+       isAdmin: true,
+};
+
+// Создать аналогичные друг другу интерфейс и тип: объект с обязательным неизменяемым свойством id типа число, обязательным 
+// полем name типа строка и опциональным полем gender с типом либо 'male' либо 'female'
+type Guest1 = { readonly id: number, name: string, gender?: "male" | "female" };
+interface Guest2 {
+       readonly id: number,
+       name: string,
+       gender?: "male" | "female"
+};
+
+// Затипизировать входящие параметры "x" и "y" как числа и возвращаемое значение
+const sumTwoNumbers = (x: number, y: number): number => x + y;
+console.log(sumTwoNumbers(7, 3));
+
+// Затипизировать входящие параметры "x", "y" и "z" как числа. Параметр "z" должен быть необязательным.
+//  Написать корректную реализацию функции с учетом необязательности "z"
+const sumThreeNumbers = (x: number, y: number, z?: number): number => {
+       if (z) {
+              return x + y + z
+       } else {
+              return x + y
+       }
+};
+console.log(sumThreeNumbers(2, 3, 6))
+console.log(sumThreeNumbers(2, 3))
+
+// Написать перегрузку и реализацию функции sum такую что: 
+// 1) если на вход приходят два числа, то возвращается число 
+// 2) если на вход приходит приходит строка и число или обе строки, то возвращается строка
+function sum(x: number, y: number): number;
+function sum(x: string, y: number): string;
+function sum(x: number, y: string): string;
+function sum(x: string, y: string): string;
+
+function sum(x: number | string, y: number | string): number | string {
+       if (typeof x === 'number' && typeof y === 'number') {
+              return x + y;
+       } else {
+              return (x.toString() + y.toString());
+       }
+
+}
+console.log(sum(11, 3))
+console.log(sum(11, "3"))
+console.log(sum("12", "3"))
+
+
+// Затипизировать this
+function getName(this: { [id: string]: { name: string } }, id: string) {
+       return this[id]?.name;
+}
+
+
+// Затипизировать возвращаемое значение
+const sayHi = (): void => {
+       console.log('hi');
+}
+
+// Тип значение которого может быть тип Book или Car
+interface Book { }
+interface Car { }
+type BookOrCar = Book | Car;
+
+// Используя типы House, City, Country создать новый тип FullAddress который включает все свойства вышеперечисленных
+interface House { apartment: number }
+interface City { zipCode: number }
+interface Country { name: string }
+type FullAddress = House & City & Country;
+const address: FullAddress = {
+       apartment: 15,
+       zipCode: 4,
+       name: "Minsk"
+}
+
+// переменная с типом строка, начинающаяся с префикса "id:" и дальше числовое id. Например, let myId = 'id:2'
+type ID = `${"id:"}${number}`
```

Comment:
тут кавычки избыточны, все что обернуто в шаблонную строку по умолчанию считается строкой
```suggestion
type ID = `id:${number}`
```